certgen/certgen.go

```diff
@@ -0,0 +1,164 @@
+// certificates.
+type CertificateConfig struct {
+	// Required fields
+	//
+	// What these are set to, including the zero values if left unset, is what
+	// will appear in the leaf certificates.
+	SubjectCommonName         string
+	SubjectOrganization       string
+	SubjectOrganizationalUnit string
+	SubjectLocality           string
+	SubjectCountry            string
+	SignatureAlgorithm        x509.SignatureAlgorithm
+
+	// Optional fields
+
+	// Prefix is a prefix that will be used in conjunction with the
+	// SubjectCommonName to create a more specific DNS SAN.
+	Prefix string
+	// NotAfterInterval specifies an interval in which the NotAfter time of a
+	// certificate must fall.
+	NotAfterInterval *interval.Interval
+}
+
+// CA is a Certificate Authority that issues synthetic certificates and
+// certificate chains using its RootCert and RootKey.
+type CA struct {
+	RootCert   *x509.Certificate
+	RootKey    crypto.Signer
+	CertConfig CertificateConfig
+}
+
+// IssueCertificate creates a new leaf certificate, issued by the root specified
+// in the RootCert and RootKey fields of the CA, and configured using the
+// CertConfig in the CA.
+func (ca *CA) IssueCertificate() (*x509.Certificate, error) {
+	key, err := rsa.GenerateKey(crand.Reader, 2048)
+	if err != nil {
+		return nil, fmt.Errorf("error generating key pair: %s", err)
+	}
+
+	template, err := leafTemplate(ca.CertConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf template: %s", err)
+	}
+
+	leafDER, err := x509.CreateCertificate(crand.Reader, template, ca.RootCert, key.Public(), ca.RootKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf certificate: %s", err)
+	}
+
+	leaf, err := x509.ParseCertificate(leafDER)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing leaf certificate DER: %s", err)
+	}
+
+	return leaf, nil
+}
+
+func leafTemplate(c CertificateConfig) (*x509.Certificate, error) {
+	sn, err := randSerialNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	notAfter := randNotAfter(c.NotAfterInterval)
+
+	return &x509.Certificate{
+		SerialNumber: sn,
+		Subject: pkix.Name{
+			Country:            []string{c.SubjectCountry},
+			Organization:       []string{c.SubjectOrganization},
+			OrganizationalUnit: []string{c.SubjectOrganizationalUnit},
+			Locality:           []string{c.SubjectLocality},
+			CommonName:         c.SubjectCommonName,
+		},
+		NotBefore:          notAfter.Add(-time.Hour * 24),
+		NotAfter:           notAfter,
+		SignatureAlgorithm: c.SignatureAlgorithm,
+
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  false,
+		DNSNames:              []string{c.SubjectCommonName, extendedDNSSAN(c.Prefix, c.SubjectCommonName)},
+	}, nil
+}
+
+func randSerialNumber() (*big.Int, error) {
+	return crand.Int(crand.Reader, big.NewInt(math.MaxInt64))
```

Comment:
Not that it's necessarily a problem, but I think this has the same bug that was reported here: https://arstechnica.com/information-technology/2019/03/godaddy-apple-and-google-goof-results-in-1-million-misissued-certificates/

`math.MaxInt64` returns the max of a signed int64, which only uses 63 of the 64 bits. I think you want `math.MaxUint64` instead.

Reply:
I used math.MaxInt64 because big.NewInt() takes an int64 as its argument, I figured I couldn't pass the max value for a uint64 in as an int64 (and have it WAI), and like you said, it's not really a problem for synthetic certificates.

I've found a hackier work around that sets it to be the value of math.MaxUint64 and done that.